xdsl/transforms/canonicalize/canonicalize_riscv.py

```diff
@@ -0,0 +1,100 @@
+from math import ceil, log2
+
+from xdsl.dialects import builtin, riscv
+from xdsl.ir import SSAValue
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.transforms.canonicalize import is_canonicalization
+
+_HAS_IMMEDIATE_FORM_OP: dict[type[riscv.RdRsRsIntegerOperation], type[riscv.RdRsImmIntegerOperation]] = {  # type: ignore[reportGeneralTypeIssues]
+    riscv.AddOp: riscv.AddiOp,
+    riscv.AndOp: riscv.AndiOp,
+    riscv.SltOp: riscv.SltiOp,
+    riscv.SllOp: riscv.SlliOp,
+    riscv.SrlOp: riscv.SrliOp,
+    riscv.SraOp: riscv.SraiOp,
+    riscv.OrOp: riscv.OriOp,
+    riscv.XorOp: riscv.XoriOp,
+    riscv.SltuOp: riscv.SltiuOp,
+}
+
+# TODO: make interface
+_COMMUTATIVE_OPS = (riscv.AddOp, riscv.AndOp, riscv.OrOp, riscv.XorOp)
+
+
+def twos_complement_bitwidth(num: int):
+    """
+    Quick approximation of number of bits required to represent a number in two's
+    complement.
+    """
+    is_neg = 1 if num < 0 else 0
+    num = abs(num)
+    return ceil(log2(num)) + is_neg
```

Comment:
```suggestion
    return num.bit_length() + is_neg
```
;)

Reply:
Python doc [link](https://docs.python.org/3/library/stdtypes.html#int.bit_length), looks like `1 + int(log(abs(x), 2))` for non-zero `x`.